myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java

```diff
@@ -74,20 +76,35 @@ public void onEvent(ResourceOffersEvent event, long sequence,
     SchedulerDriver driver = event.getDriver();
     List<Offer> offers = event.getOffers();
 
+    // Sometimes, we see that mesos sends resource offers before Myriad receives
+    // a notification for "framework registration". This is a simple defensive code
+    // to not process any offers unless Myriad receives a "framework registered" notification.
+    if (schedulerState.getFrameworkID() == null) {
+      LOGGER.warn("Received {} offers, but declining them since Framework ID is not yet set", offers.size());
+      for (Offer offer : offers) {
+        driver.declineOffer(offer.getId());
+      }
+      return;
+    }
     LOGGER.info("Received offers {}", offers.size());
     LOGGER.debug("Pending tasks: {}", this.schedulerState.getPendingTaskIds());
     driverOperationLock.lock();
     try {
       for (Iterator<Offer> iterator = offers.iterator(); iterator.hasNext();) {
         Offer offer = iterator.next();
+        NodeTask nodeTask = schedulerState.getNodeTask(offer.getSlaveId());
```

Comment:
@yufeldman I realized if we update the slave attributes only during launchTasks, we may lose the slave attributes when RM fails over. This is because slave attributes in NodeTask are currently not being persisted.

We have a couple of options:
- Update slave attributes on every offer (like we are doing it now).
- Persist slave attributes with NodeTask. Implement attribute change notification (in the future) and update NodeTask when slave attributes change.